fog/view/server/src/fog_view_router_request_handler.rs

```diff
@@ -0,0 +1,226 @@
+        if shard_clients.is_empty() {
+            break;
+        }
+
+        authenticate_view_stores(
+            enclave.clone(),
+            processed_shard_response_data.view_store_uris_for_authentication,
+            logger.clone(),
+        )
+        .await?;
+    }
+
+    // TODO: Collate the query_responses into one response for the client. Make an
+    // enclave  method for this.
+    let response = FogViewRouterResponse::new();
+    Ok(response)
+}
+
+/// Sends a client's query request to all of the Fog View shards.
+async fn route_query(
+    request: &MultiViewStoreQueryRequest,
+    shard_clients: Vec<Arc<FogViewApiClient>>,
+) -> Result<Vec<(Arc<FogViewApiClient>, MultiViewStoreQueryResponse)>, RouterServerError> {
+    let responses = shard_clients
+        .into_iter()
+        .map(|shard_client| query_shard(request, shard_client.clone()));
+    try_join_all(responses).await
+}
+
+/// Sends a client's query request to one of the Fog View shards.
+async fn query_shard(
+    request: &MultiViewStoreQueryRequest,
+    shard_client: Arc<FogViewApiClient>,
+) -> Result<(Arc<FogViewApiClient>, MultiViewStoreQueryResponse), RouterServerError> {
+    let client_unary_receiver = shard_client.multi_view_store_query_async(request)?;
+    let response = client_unary_receiver.await?;
+
+    Ok((shard_client, response))
+}
+
+/// Authenticates Fog View Stores that have previously not been authenticated.
+pub async fn authenticate_view_stores<E: ViewEnclaveProxy>(
+    enclave: E,
+    view_store_uris: Vec<FogViewUri>,
+    logger: Logger,
+) -> Result<Vec<()>, RpcStatus> {
+    let pending_auth_requests = view_store_uris
+        .into_iter()
+        .map(|store_uri| authenticate_view_store(enclave.clone(), store_uri, logger.clone()));
+
+    try_join_all(pending_auth_requests).await.map_err(|err| {
+        router_server_err_to_rpc_status(
+            "Query: cannot authenticate with each Fog View Store:",
+            err,
+            logger.clone(),
+        )
+    })
+}
+
+/// Authenticates a Fog View Store that has previously not been authenticated.
+pub async fn authenticate_view_store<E: ViewEnclaveProxy>(
+    enclave: E,
+    view_store_url: FogViewUri,
+    logger: Logger,
+) -> Result<(), RouterServerError> {
+    let view_store_id = ResponderId::from_str(&view_store_url.to_string())?;
+    let client_auth_request = enclave.view_store_init(view_store_id.clone())?;
+    let grpc_env = Arc::new(
+        grpcio::EnvBuilder::new()
+            .name_prefix("Main-RPC".to_string())
+            .build(),
+    );
+    let view_store_client = FogViewApiClient::new(
+        ChannelBuilder::default_channel_builder(grpc_env).connect_to_uri(&view_store_url, &logger),
+    );
+
+    let auth_unary_receiver = view_store_client.auth_async(&client_auth_request.into())?;
+    let auth_response = auth_unary_receiver.await?;
+
+    let result = enclave.view_store_connect(view_store_id, auth_response.into())?;
+
+    Ok(result)
```

Comment:
I try to not put everything in one line because it makes debugging harder. If I want to know the value of the result while using a debugging breakpoint, I can't do that easily with the second way.